util/compression.h

```diff
@@ -796,4 +799,42 @@ inline char* ZSTD_Uncompress(const char* input_data, size_t input_length,
   return nullptr;
 }
 
+inline std::string ZSTD_TrainDictionary(const std::string& samples,
+                                        const std::vector<size_t>& sample_lens,
+                                        size_t max_dict_bytes) {
+  // Dictionary trainer is available since v0.6.1, but ZSTD was marked stable
+  // only since v0.8.0. For now we enable the feature in stable versions only.
+#if ZSTD_VERSION_NUMBER >= 800  // v0.8.0+
+  std::unique_ptr<char[]> dict_data(new char[max_dict_bytes]);
+  size_t dict_len =
+      ZDICT_trainFromBuffer(&dict_data[0], max_dict_bytes, &samples[0],
+                            &sample_lens[0], sample_lens.size());
+  if (ZDICT_isError(dict_len)) {
+    return "";
+  }
+  // TODO(ajkr): we may be able to avoid this copy by storing
+  // compression_dict in something other than std::string.
+  return std::string(&dict_data[0], dict_len);
```

Comment:
According to this answer the following should be safe:
https://stackoverflow.com/questions/39200665/directly-write-into-char-buffer-of-stdstring
```
std::string dict_data(max_dict_bytes, '\0');
size_t dict_len =
      ZDICT_trainFromBuffer(&dict_data[0], max_dict_bytes, &samples[0],
                            &sample_lens[0], sample_lens.size());
...
return dict_data;
```

Reply:
Thanks, Yi.